File: app/frontend/src/core/services/types/MediaProcessing.ts
import { ServiceError } from '../base/types';
import { Readable } from 'stream';

export enum ProcessingJobStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled'
}

export interface ProcessingProgress {
  thumbnail: number;
  conversion: number;
  metadata: number;
  overall: number;
  total: number;
}

export interface ProcessingStats {
  totalJobs: number;
  completedJobs: number;
  failedJobs: number;
  averageProcessingTime: number;
  peakMemoryUsage: number;
  currentMemoryUsage: number;
}

export interface ThumbnailOptions {
  format?: string;
  width?: number;
  height?: number;
  quality?: number;
  fit?: 'cover' | 'contain' | 'fill' | 'inside' | 'outside';
  position?: 'top' | 'right top' | 'right' | 'right bottom' | 'bottom' | 'left bottom' | 'left' | 'left top' | 'center';
  background?: string;
  progressive?: boolean;
  withMetadata?: boolean;
}

export interface ThumbnailResult {
  data: Buffer;
  format: string;
  width: number;
  height: number;
  size: number;
}

export interface ConversionOptions {
  targetFormat: string;
  quality?: number;
  width?: number;
  height?: number;
  fps?: number;
  bitrate?: string;
  codec?: string;
  preset?: string;
  metadata?: Record<string, string>;
}

export interface ConversionResult {
  data: Buffer;
  format: string;
  size: number;
  duration?: number;
  width?: number;
  height?: number;
  fps?: number;
  bitrate?: string;
  codec?: string;
}

export interface MetadataExtractionOptions {
  type?: string;
  includeExif?: boolean;
  includeIptc?: boolean;
  includeXmp?: boolean;
  includeIcc?: boolean;
  raw?: boolean;
}

export interface MetadataExtractionResult {
  width?: number;
  height?: number;
  duration?: number;
  bitrate?: number;
  codec?: string;
  fps?: number;
  rotation?: number;
  createdAt?: Date;
  modifiedAt?: Date;
  exif?: Record<string, any>;
  iptc?: Record<string, any>;
  xmp?: Record<string, any>;
  icc?: Record<string, any>;
  raw?: Record<string, any>;
}

export interface MetadataResult {
  format: string;
  size: number;
  width?: number;
  height?: number;
  duration?: number;
  bitrate?: number;
  codec?: string;
  fps?: number;
  rotation?: number;
  createdAt?: Date;
  modifiedAt?: Date;
  exif?: Record<string, any>;
  iptc?: Record<string, any>;
  xmp?: Record<string, any>;
  icc?: Record<string, any>;
  raw?: Record<string, any>;
}

export interface MediaProcessingOptions {
  thumbnail?: ThumbnailOptions;
  conversion?: ConversionOptions;
  metadata?: MetadataExtractionOptions;
}

export interface MediaProcessingResult {
  processingTime: number;
  thumbnail?: ThumbnailResult;
  conversion?: ConversionResult;
  metadata?: MetadataResult;
}

export interface ProcessingJob {
  id: string;
  input: Buffer | string;
  options: MediaProcessingOptions;
  status: ProcessingJobStatus;
  progress: ProcessingProgress;
  startTime: number;
  endTime?: number;
  processingTime?: number;
  error?: ServiceError;
  result?: MediaProcessingResult;
}

export interface StreamProcessingOptions extends MediaProcessingOptions {
  chunkSize?: number;
  maxParallel?: number;
  expectedSize?: number;
  conversion: ConversionOptions;
}

export interface StreamProcessingResult extends MediaProcessingResult {
  bytesProcessed: number;
  totalBytes: number;
  originalFormat: string;
  memoryUsage?: {
    peak: number;
    final: number;
  };
}

export interface ProcessingPlugin {
  id: string;
  name: string;
  version: string;
  type: string;
  initialize(): Promise<void>;
  process(input: Buffer | string, options: any): Promise<any>;
  cleanup(): Promise<void>;
  getSupportedFormats(): string[];
}

export interface PluginManager {
  loadPlugins(directory: string): Promise<ProcessingPlugin[]>;
  unloadPlugin(pluginId: string): Promise<void>;
  getLoadedPlugins(): Promise<ProcessingPlugin[]>;
  getPlugins(): ProcessingPlugin[];
} 

File: backend/app/src/core/services/types/MediaProcessing.py
from typing import Any, Dict, List, Union
from enum import Enum


class ProcessingJobStatus(Enum):
    PENDING = 'pending'
    PROCESSING = 'processing'
    COMPLETED = 'completed'
    FAILED = 'failed'
    CANCELLED = 'cancelled'
class ProcessingProgress:
    thumbnail: float
    conversion: float
    metadata: float
    overall: float
    total: float
class ProcessingStats:
    totalJobs: float
    completedJobs: float
    failedJobs: float
    averageProcessingTime: float
    peakMemoryUsage: float
    currentMemoryUsage: float
class ThumbnailOptions:
    format?: str
    width?: float
    height?: float
    quality?: float
    fit?: Union['cover', 'contain', 'fill', 'inside', 'outside']
    position?: Union['top', 'right top', 'right', 'right bottom', 'bottom', 'left bottom', 'left', 'left top', 'center']
    background?: str
    progressive?: bool
    withMetadata?: bool
class ThumbnailResult:
    data: Buffer
    format: str
    width: float
    height: float
    size: float
class ConversionOptions:
    targetFormat: str
    quality?: float
    width?: float
    height?: float
    fps?: float
    bitrate?: str
    codec?: str
    preset?: str
    metadata?: Dict[str, str>
class ConversionResult:
    data: Buffer
    format: str
    size: float
    duration?: float
    width?: float
    height?: float
    fps?: float
    bitrate?: str
    codec?: str
class MetadataExtractionOptions:
    type?: str
    includeExif?: bool
    includeIptc?: bool
    includeXmp?: bool
    includeIcc?: bool
    raw?: bool
class MetadataExtractionResult:
    width?: float
    height?: float
    duration?: float
    bitrate?: float
    codec?: str
    fps?: float
    rotation?: float
    createdAt?: Date
    modifiedAt?: Date
    exif?: Dict[str, Any>
    iptc?: Dict[str, Any>
    xmp?: Dict[str, Any>
    icc?: Dict[str, Any>
    raw?: Dict[str, Any>
class MetadataResult:
    format: str
    size: float
    width?: float
    height?: float
    duration?: float
    bitrate?: float
    codec?: str
    fps?: float
    rotation?: float
    createdAt?: Date
    modifiedAt?: Date
    exif?: Dict[str, Any>
    iptc?: Dict[str, Any>
    xmp?: Dict[str, Any>
    icc?: Dict[str, Any>
    raw?: Dict[str, Any>
class MediaProcessingOptions:
    thumbnail?: \'ThumbnailOptions\'
    conversion?: \'ConversionOptions\'
    metadata?: \'MetadataExtractionOptions\'
class MediaProcessingResult:
    processingTime: float
    thumbnail?: \'ThumbnailResult\'
    conversion?: \'ConversionResult\'
    metadata?: \'MetadataResult\'
class ProcessingJob:
    id: str
    input: Union[Buffer, str]
    options: \'MediaProcessingOptions\'
    status: \'ProcessingJobStatus\'
    progress: \'ProcessingProgress\'
    startTime: float
    endTime?: float
    processingTime?: float
    error?: ServiceError
    result?: \'MediaProcessingResult\'
class StreamProcessingOptions:
    chunkSize?: float
    maxParallel?: float
    expectedSize?: float
    conversion: \'ConversionOptions\'
class StreamProcessingResult:
    bytesProcessed: float
    totalBytes: float
    originalFormat: str
    memoryUsage?: {
    peak: float
    final: float
}
class ProcessingPlugin:
    id: str
    name: str
    version: str
    type: str
    initialize(): Awaitable[None>
    process(input: Union[Buffer, str, options: Any): Awaitable[Any>]
    cleanup(): Awaitable[None>
    getSupportedFormats(): List[str]
class PluginManager:
    loadPlugins(directory: str): Awaitable[ProcessingPlugin[]>
    unloadPlugin(pluginId: str): Awaitable[None>
    getLoadedPlugins(): Awaitable[ProcessingPlugin[]>
    getPlugins(): List[ProcessingPlugin] 